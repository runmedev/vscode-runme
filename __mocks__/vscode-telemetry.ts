import { vi } from 'vitest'

export const TelemetryReporter = {
  sendTelemetryEvent: vi.fn(),
  sendTelemetryErrorEvent: vi.fn()
}

export class TelemetryViewProvider {
  constructor() {
  }
}
