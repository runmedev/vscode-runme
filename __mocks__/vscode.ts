import path from 'node:path'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'

import { vi } from 'vitest'
import { URI } from 'vscode-uri'
import type { Command, Disposable } from 'vscode'


export const notebooks = {
  createNotebookController: vi.fn().mockReturnValue({
    dispose: vi.fn(),
    createNotebookCellExecution: vi.fn().mockReturnValue({ start: vi.fn(), end: vi.fn() })
  }),
  registerNotebookCellStatusBarItemProvider: vi.fn(),
  createRendererMessaging: vi.fn().mockReturnValue({
    postMessage: vi.fn(),
    onDidReceiveMessage: vi.fn().mockReturnValue({ dispose: vi.fn() })
  }),
}

export class Uri extends URI {
  static file = vi.fn(super.file)
  static parse = vi.fn(super.parse)

  static joinPath = vi.fn((uri: Uri, ...paths: string[]) => {
    return super.file(path.join(uri.fsPath, ...paths))
  })
}

export const workspace = {
  getConfiguration: vi.fn().mockReturnValue(new Map()),
  onDidChangeConfiguration: vi.fn(),
  openNotebookDocument: vi.fn().mockReturnValue({ uri: 'new notebook uri' }),
  openTextDocument: vi.fn().mockReturnValue({
    getText: vi.fn().mockReturnValue(readFileSync(join(__dirname, 'gitignore.mock'), 'utf8'))
  }),
  registerNotebookSerializer: vi.fn(),
  onDidOpenNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  onDidSaveNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  onDidChangeNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  fs: {
    writeFile: vi.fn(),
    readFile: vi.fn().mockResolvedValue(Buffer.from('some wasm file')),
    stat: vi.fn().mockResolvedValue(1),
    createDirectory: vi.fn(),
    delete: vi.fn()
  },
  createFileSystemWatcher: vi.fn().mockReturnValue({
    onDidCreate: vi.fn(),
    onDidDelete: vi.fn()
  }),
  workspaceFolders: [ { uri: { fsPath: '/runme/workspace' } } ],
  findFiles: vi.fn().mockReturnValue([
    {
      path: 'runme/workspace/README.md',
    },
    {
      path: 'runme/workspace/src/README.md',
    },
    {
      path: 'runme/workspace/src/COMMANDS.md',
    },
    {
      path: 'runme/workspace/src/RUNME.md',
    }
  ]),
}

export const terminal = {
  show: vi.fn(),
  sendText: vi.fn(),
  dispose: vi.fn()
}

export const window = {
  createOutputChannel: vi.fn().mockReturnValue({
    appendLine: vi.fn()
  }),
  showWarningMessage: vi.fn(),
  showInformationMessage: vi.fn(),
  showErrorMessage: vi.fn(),
  createTerminal: vi.fn().mockReturnValue(terminal),
  showNotebookDocument: vi.fn(),
  showTextDocument: vi.fn(),
  onDidChangeActiveNotebookEditor: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  registerTreeDataProvider: vi.fn(),
  registerUriHandler: vi.fn(),
  onDidCloseTerminal: vi.fn(),
  withProgress: vi.fn(),
  onDidChangeActiveColorTheme: vi.fn().mockReturnValue({ dispose: vi.fn() }),
}

export const tasks = {
  registerTaskProvider: vi.fn(),
  onDidEndTaskProcess: vi.fn(),
  onDidStartTaskProcess: vi.fn(),
  executeTask: vi.fn().mockResolvedValue(undefined)
}

export const commands = {
  registerCommand: vi.fn(),
  executeCommand: vi.fn(),
  getCommands: vi.fn().mockResolvedValue(['_notebook.selectKernel'])
}

export enum ViewColumn {
  Beside = 'Beside'
}

export const env = {
  clipboard: {
    writeText: vi.fn()
  },
  openExternal: vi.fn()
}

export const NotebookData = vi.fn()
export const NotebookCellData = vi.fn()
export enum NotebookCellKind {
  Code = 1,
  Markup = 2
}

export const TreeItem = vi.fn()
export const Event = vi.fn()

export enum TreeItemCollapsibleState {
  None = 0,
  Collapsed = 1,
  Expanded = 2
}

export enum FileType {
  /**
   * The file type is unknown.
   */
  Unknown = 0,
  /**
   * A regular file.
   */
  File = 1,
  /**
   * A directory.
   */
  Directory = 2,
  /**
   * A symbolic link to a file.
   */
  SymbolicLink = 64
}

export enum NotebookCellStatusBarAlignment {
  Left = 1,
  Right = 2
}

export class NotebookCellStatusBarItem {
  label: string

  alignment: NotebookCellStatusBarAlignment

  tooltip?: string

  priority?: number

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(label: string, alignment: NotebookCellStatusBarAlignment) {
    this.label = label
    this.alignment = alignment
  }
}



export class NotebookCellOutputItem {

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  static json(value: any, mime = 'text/plain'): NotebookCellOutputItem {
    return {
      mime: mime,
      data: new Uint8Array()
    }
  }

  mime: string
  data: Uint8Array
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(data: Uint8Array, mime: string) {}
}

export class NotebookCellOutput {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(public items: NotebookCellOutputItem[], public metadata?: { [key: string]: any }) {}
}

export const ProgressLocation = {
  Window: 1
}

type MessageCallback<T> = (message: T) => void
type Event<T> = (listener: MessageCallback<T>) => Disposable

export class EventEmitter<T> {
  listeners: MessageCallback<T>[] = []

  event: Event<T> = (listener) => {
    this.listeners.push(listener)

    return {
      dispose: () => {
        this.listeners = this.listeners.filter(x => x !== listener)
      }
    }
  }

  fire(data: T) {
    this.listeners.forEach(l => l(data))
  }

  async fireAsync(data: T) {
    await Promise.all(
      this.listeners.map(l => l(data))
    )
  }

  dispose() { }
}

export const languages = {
  registerCodeLensProvider: vi.fn(),
}

export class Position {
  constructor(
    public line: number,
    public character: number
  ) { }

  with(line?: number, character?: number): Position {
    return new Position(
      line ?? this.line,
      character ?? this.character,
    )
  }
}

export class Range {
  constructor(
    public start: Position,
    public end: Position,
  ) { }
}

export class CodeLens {
  constructor(
    public range: Range,
    public command?: Command
  ) { }
}

export enum TaskScope {
  /**
   * The task is a global task. Global tasks are currently not supported.
   */
  Global = 1,

  /**
   * The task is a workspace task
   */
  Workspace = 2
}

export const ShellExecution = vi.fn()
export const Task = vi.fn()
