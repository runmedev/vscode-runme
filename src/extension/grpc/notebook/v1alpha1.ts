export * from '@buf/stateful_runme.community_timostamm-protobuf-ts/runme/notebook/v1alpha1/notebook_pb'
