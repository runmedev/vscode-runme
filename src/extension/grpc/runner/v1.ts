export * from '@buf/stateful_runme.community_timostamm-protobuf-ts/runme/runner/v1/runner_pb'
