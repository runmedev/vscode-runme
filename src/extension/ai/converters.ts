// Conversion routines between the protocol buffer format and VSCode's representation of a notebook
//
// See ../vscode_apis.md for an exlanation. It is very helpful for understanding this folder.

import * as vscode from 'vscode'
import * as parser_pb from '@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb'

import { ServerLifecycleIdentity, getServerConfigurationValue } from '../../utils/configuration'
import { Serializer } from '../../types'
import * as serializerTypes from '../grpc/serializerTypes'
import * as serializer from '../serializer'
import { Kernel } from '../kernel'

import * as protos from './protos'

// Converter provides converstion routines from vscode data types to protocol buffer types.
// It is a class because in order to handle the conversion we need to keep track of the kernel
// because we need to add execution information to the cells before serializing.
export class Converter {
  kernel: Kernel
  constructor(kernel: Kernel) {
    this.kernel = kernel
  }

  // notebokDataToProto converts a VSCode NotebookData to a RunMe Notebook proto.
  // It adds execution information to the cells before converting.
  public async notebookDataToProto(notebookData: vscode.NotebookData): Promise<parser_pb.Notebook> {
    // We need to add the execution info to the cells so that the AI model can use that information.
    const cellDataWithExec = await serializer.SerializerBase.addExecInfo(notebookData, this.kernel)
    let notebookDataWithExec = new vscode.NotebookData(cellDataWithExec)
    // marshalNotebook returns a protocol buffer using the ts client library from buf we need to
    // convert it to es
    let notebookProto = serializer.GrpcSerializer.marshalNotebook(notebookDataWithExec)
    return protos.notebookTSToES(notebookProto)
  }
}

// cellToCellData converts a NotebookCell to a NotebookCellData.
// NotebookCell is an interface used by the editor.
// NotebookCellData is a concrete class.
// For more explanation of the different vscode APIs see
// https://github.com/jlewi/foyle/blob/main/frontend/foyle/vscode_apis.md
export function cellToCellData(cell: vscode.NotebookCell): vscode.NotebookCellData {
  let data = new vscode.NotebookCellData(
    cell.kind,
    cell.document.getText(),
    cell.document.languageId,
  )

  data.metadata = cell.metadata
  data.outputs = []
  for (let o of cell.outputs) {
    data.outputs.push(o)
  }
  return data
}

// cellProtosToCellData converts an array of RunMe cell protos to an array of VSCode CellData
export function cellProtosToCellData(cells: serializerTypes.Cell[]): vscode.NotebookCellData[] {
  let notebook: Serializer.Notebook = {
    cells: [],
  }
  for (let cell of cells) {
    let kind: vscode.NotebookCellKind = vscode.NotebookCellKind.Markup

    if (cell.kind === serializerTypes.CellKind.CODE) {
      kind = vscode.NotebookCellKind.Code
    }

    let newCell: Serializer.Cell = {
      value: cell.value,
      metadata: cell.metadata,
      kind: kind,
      languageId: cell.languageId,
      // TODO(jeremy): Should we include outputs? The generate response should never contain outputs so we shouldn't
      // have to worry about them.
    }
    notebook.cells.push(newCell)
  }

  const identity: ServerLifecycleIdentity = getServerConfigurationValue<ServerLifecycleIdentity>(
    'lifecycleIdentity',
    serializerTypes.RunmeIdentity.ALL,
  )
  let newCellData = serializer.SerializerBase.revive(notebook, identity)
  return newCellData
}

// vsCellsToESProto converts a VSCode NotebookCell to a RunMe Cell proto
// Its not quite the inverse of cellProtosToCellData because this function
// uses the ES client library as opposed to the TS client library.
// The reason we don't rely on the serialization routines is we don't want to
// generate an RPC just to convert the cell to a proto.
export function vsCellsToESProtos(cells: vscode.NotebookCell[]): parser_pb.Cell[] {
  const cellProtos: parser_pb.Cell[] = []

  for (let cell of cells) {
    const cellProto = new parser_pb.Cell()
    if (cell.kind === vscode.NotebookCellKind.Code) {
      cellProto.kind = parser_pb.CellKind.CODE
    } else if (cell.kind === vscode.NotebookCellKind.Markup) {
      cellProto.kind = parser_pb.CellKind.MARKUP
    }

    cellProto.value = cell.document.getText()
    cellProto.metadata = cell.metadata

    cellProtos.push(cellProto)
  }

  return cellProtos
}
