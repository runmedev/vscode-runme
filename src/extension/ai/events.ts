import { PromiseClient } from '@connectrpc/connect'
import { AIService } from '@buf/jlewi_foyle.connectrpc_es/foyle/v1alpha1/agent_connect'
import {
  LogEventsRequest,
  LogEventType,
  LogEvent,
  LogEvent_ExecuteStatus,
} from '@buf/jlewi_foyle.bufbuild_es/foyle/v1alpha1/agent_pb'
import * as vscode from 'vscode'
import { ulid } from 'ulidx'

import { RUNME_CELL_ID } from '../constants'
import getLogger from '../logger'

import { SessionManager } from './sessions'
import * as converters from './converters'

// Interface for the event reporter
// This allows us to swap in a null op logger when AI isn't enabled
export interface IEventReporter {
  reportExecution(cell: vscode.NotebookCell, status: boolean): Promise<void>
  reportEvents(events: LogEvent[]): Promise<void>
}

// EventReporter handles reporting events to the AI service
export class EventReporter implements IEventReporter {
  client: PromiseClient<typeof AIService>
  log: ReturnType<typeof getLogger>

  constructor(client: PromiseClient<typeof AIService>) {
    this.client = client
    this.log = getLogger('AIEventReporter')
  }

  async reportExecution(cell: vscode.NotebookCell, executionSuccess: boolean) {
    const contextCells: vscode.NotebookCell[] = []

    // Include some previous cells as context.
    // N.B. In principle we shouldn't need to send any additional context because we
    // set the context id. So as soon as we put the focus on the execution cell we should
    // start a streaming generate request which will include the entire notebook or a large portion of it.
    // However, we still send some additional context here for two reasons
    // 1. Help us verify that sending context ids is working correctly.
    // 2. Its possible in the future we start rate limiting streaming generate requests and don't want to rely on it
    // for providing the context of the cell execution.
    let startIndex = cell.index - 1
    if (startIndex < 0) {
      startIndex = 0
    }
    for (let i = startIndex; i < cell.index; i++) {
      contextCells.push(cell.notebook.cellAt(i))
    }

    contextCells.push(cell)
    const cells = converters.vsCellsToESProtos(contextCells)
    const event = new LogEvent()
    event.selectedId = cell.metadata?.[RUNME_CELL_ID]
    event.selectedIndex = cell.index
    event.type = LogEventType.EXECUTE
    event.cells = cells
    event.contextId = SessionManager.getManager().getID()

    if (executionSuccess) {
      event.executeStatus = LogEvent_ExecuteStatus.SUCCEEDED
    } else {
      event.executeStatus = LogEvent_ExecuteStatus.FAILED
    }
    return this.reportEvents([event])
  }

  async reportEvents(events: LogEvent[]) {
    const req = new LogEventsRequest()
    req.events = events
    for (const event of events) {
      if (event.eventId === '') {
        event.eventId = ulid()
      }
    }
    await this.client.logEvents(req).catch((e) => {
      this.log.error(`Failed to log event; error: ${e}`)
    })
  }
}

// NullOpEventReporter is a null op implementation of the event reporter
export class NullOpEventReporter implements IEventReporter {
  async reportExecution(_cell: vscode.NotebookCell, _status: boolean) {
    // Do nothing
  }

  async reportEvents(_events: LogEvent[]) {
    // Do nothing
  }
}

let _globalReporter = new NullOpEventReporter()

// getEventReporter returns the global event reporter
export function getEventReporter(): IEventReporter {
  return _globalReporter
}

export function setEventReporter(reporter: IEventReporter) {
  _globalReporter = reporter
}
