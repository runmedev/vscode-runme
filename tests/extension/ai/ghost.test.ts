import { vi, suite, test, expect } from 'vitest'
import { workspace } from 'vscode'
import * as agent_pb from '@buf/jlewi_foyle.bufbuild_es/foyle/v1alpha1/agent_pb'
import * as parser_pb from '@buf/stateful_runme.bufbuild_es/runme/parser/v1/parser_pb'

import { GhostCellGenerator } from '../../../src/extension/ai/ghost'
import { cellToCellData } from '../../../src/extension/ai/converters'
import { CellChangeEvent } from '../../../src/extension/ai/stream'
import { Uri } from '../../../__mocks__/vscode'

const changeEvents: [CellChangeEvent, { debounced: boolean; firstRequest: boolean }][] = [
  [
    {
      notebookUri: 'h',
      cellIndex: 0,
      trigger: 1,
    },
    { debounced: true, firstRequest: false },
  ],
  [
    {
      notebookUri: 'he',
      cellIndex: 1,
      trigger: 1,
    },
    { debounced: true, firstRequest: false },
  ],
  [
    {
      notebookUri: 'hel',
      cellIndex: 2,
      trigger: 1,
    },
    { debounced: true, firstRequest: false },
  ],
  [
    {
      notebookUri: 'hell',
      cellIndex: 3,
      trigger: 1,
    },
    { debounced: true, firstRequest: false },
  ],
  [
    {
      notebookUri: 'hello',
      cellIndex: 4,
      trigger: 1,
    },
    { debounced: false, firstRequest: true },
  ],
  [
    {
      notebookUri: 'how',
      cellIndex: 5,
      trigger: 1,
    },
    { debounced: true, firstRequest: false },
  ],
  [
    {
      notebookUri: 'how are',
      cellIndex: 6,
      trigger: 1,
    },
    { debounced: true, firstRequest: false },
  ],
  [
    {
      notebookUri: 'how are you?',
      cellIndex: 7,
      trigger: 1,
    },
    { debounced: false, firstRequest: true },
  ],
  [
    {
      notebookUri: "Is it me you're looking for?",
      cellIndex: 8,
      trigger: 1,
    },
    { debounced: false, firstRequest: true },
  ],
  [
    {
      notebookUri: 'here we go',
      cellIndex: 9,
      trigger: 1,
    },
    { debounced: false, firstRequest: true },
  ],
  [
    {
      notebookUri: 'again',
      cellIndex: 10,
      trigger: 1,
    },
    { debounced: false, firstRequest: true },
  ],
  [
    {
      notebookUri: 'down the only road I have ever known',
      cellIndex: 11,
      trigger: 1,
    },
    { debounced: false, firstRequest: true },
  ],
]

const CELL: parser_pb.Cell = new parser_pb.Cell({
  kind: 1,
  value: '',
  languageId: 'markdown',
  outputs: [],
  metadata: {},
})

const NOTEBOOK = {
  cells: [CELL],
  metadata: {},
}

vi.mock('vscode')
vi.mock('vscode-telemetry')

vi.mock('../../../src/extension/grpc/client', () => ({}))
vi.mock('../../../src/extension/runner', () => ({}))

vi.mock('../../../src/extension/ai/converters', () => ({
  cellToCellData: vi.fn(),
}))

vi.mock('ulidx', () => {
  return {
    ulid: function () {
      return '01JHRGWFAKE5FAKE8FAKEWFAKE'
    },
  }
})

suite('GhostCellGenerator', () => {
  let c: parser_pb.Cell
  vi.mocked(workspace.notebookDocuments.find).mockImplementation(() => {
    return {
      uri: Uri.parse('test.md'),
      cellAt: vi.fn().mockImplementation((index) => {
        c = new parser_pb.Cell(CELL)
        const [event] = changeEvents[index]
        c.value = event.notebookUri
        return c
      }),
      getCells: vi.fn().mockReturnValue([c]),
    } as any
  })
  vi.mocked(cellToCellData).mockReturnValue({} as any)

  const Converter = class {
    notebookDataToProto = async () => {
      const n = { ...NOTEBOOK }
      n.cells = [c]
      return n
    }
  }

  test('Process events for debouncing', async () => {
    const gen = new GhostCellGenerator(new Converter() as any)
    const requests = await processEvents(async (event) => {
      const sequence = event.cellIndex
      const request = await gen.buildRequestForDebounce(event)
      return {
        sequence,
        request,
      }
    })

    expect(requests).toHaveLength(12)
    expect(requests).toMatchSnapshot()

    const cases = new Set(requests.map((r) => r.request.request.case))
    expect(cases).toEqual(new Set(['update']))
  })

  test('Build requests for event stream', async () => {
    const gen = new GhostCellGenerator(new Converter() as any)
    const requests = await processEvents(async (event, firstRequest) => {
      const sequence = event.cellIndex
      const request = await gen.buildRequest(event, firstRequest)
      return {
        sequence,
        request,
      }
    })

    expect(requests).toHaveLength(12)
    expect(requests).toMatchSnapshot()

    const cases = new Set(requests.map((r) => r.request.request.case))
    expect(cases).toEqual(new Set(['update', 'fullContext']))
  })
})

type TestEventSequence = {
  sequence: number
  request: agent_pb.StreamGenerateRequest
}

async function processEvents(
  process: (event: CellChangeEvent, firstRequest: boolean) => Promise<TestEventSequence>,
): Promise<Array<TestEventSequence>> {
  const requests = new Array<TestEventSequence>()

  for (const [event, { firstRequest }] of changeEvents) {
    const r = await process(event, firstRequest)

    if (r) {
      requests.push(r)
    }
  }
  return requests
}
