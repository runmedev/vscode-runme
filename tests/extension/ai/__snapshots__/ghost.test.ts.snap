// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GhostCellGenerator > Build requests for event stream 1`] = `
[
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "h",
        },
      },
    },
    "sequence": 0,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "he",
        },
      },
    },
    "sequence": 1,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "hel",
        },
      },
    },
    "sequence": 2,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "hell",
        },
      },
    },
    "sequence": 3,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "fullContext": {
        "notebook": {
          "cells": [
            {
              "kind": "CELL_KIND_MARKUP",
              "languageId": "markdown",
              "metadata": {},
              "outputs": [],
              "value": "hello",
            },
          ],
          "metadata": {},
        },
        "notebookUri": "file:///test.md",
        "selected": 0,
      },
      "trigger": "CELL_TEXT_CHANGE",
    },
    "sequence": 4,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "how",
        },
      },
    },
    "sequence": 5,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "how are",
        },
      },
    },
    "sequence": 6,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "fullContext": {
        "notebook": {
          "cells": [
            {
              "kind": "CELL_KIND_MARKUP",
              "languageId": "markdown",
              "metadata": {},
              "outputs": [],
              "value": "how are you?",
            },
          ],
          "metadata": {},
        },
        "notebookUri": "file:///test.md",
        "selected": 0,
      },
      "trigger": "CELL_TEXT_CHANGE",
    },
    "sequence": 7,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "fullContext": {
        "notebook": {
          "cells": [
            {
              "kind": "CELL_KIND_MARKUP",
              "languageId": "markdown",
              "metadata": {},
              "outputs": [],
              "value": "Is it me you're looking for?",
            },
          ],
          "metadata": {},
        },
        "notebookUri": "file:///test.md",
        "selected": 0,
      },
      "trigger": "CELL_TEXT_CHANGE",
    },
    "sequence": 8,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "fullContext": {
        "notebook": {
          "cells": [
            {
              "kind": "CELL_KIND_MARKUP",
              "languageId": "markdown",
              "metadata": {},
              "outputs": [],
              "value": "here we go",
            },
          ],
          "metadata": {},
        },
        "notebookUri": "file:///test.md",
        "selected": 0,
      },
      "trigger": "CELL_TEXT_CHANGE",
    },
    "sequence": 9,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "fullContext": {
        "notebook": {
          "cells": [
            {
              "kind": "CELL_KIND_MARKUP",
              "languageId": "markdown",
              "metadata": {},
              "outputs": [],
              "value": "again",
            },
          ],
          "metadata": {},
        },
        "notebookUri": "file:///test.md",
        "selected": 0,
      },
      "trigger": "CELL_TEXT_CHANGE",
    },
    "sequence": 10,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "fullContext": {
        "notebook": {
          "cells": [
            {
              "kind": "CELL_KIND_MARKUP",
              "languageId": "markdown",
              "metadata": {},
              "outputs": [],
              "value": "down the only road I have ever known",
            },
          ],
          "metadata": {},
        },
        "notebookUri": "file:///test.md",
        "selected": 0,
      },
      "trigger": "CELL_TEXT_CHANGE",
    },
    "sequence": 11,
  },
]
`;

exports[`GhostCellGenerator > Process events for debouncing 1`] = `
[
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "h",
        },
      },
    },
    "sequence": 0,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "he",
        },
      },
    },
    "sequence": 1,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "hel",
        },
      },
    },
    "sequence": 2,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "hell",
        },
      },
    },
    "sequence": 3,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "hello",
        },
      },
    },
    "sequence": 4,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "how",
        },
      },
    },
    "sequence": 5,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "how are",
        },
      },
    },
    "sequence": 6,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "how are you?",
        },
      },
    },
    "sequence": 7,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "Is it me you're looking for?",
        },
      },
    },
    "sequence": 8,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "here we go",
        },
      },
    },
    "sequence": 9,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "again",
        },
      },
    },
    "sequence": 10,
  },
  {
    "request": {
      "contextId": "01JHRGWFAKE5FAKE8FAKEWFAKE",
      "trigger": "CELL_TEXT_CHANGE",
      "update": {
        "cell": {
          "kind": "CELL_KIND_MARKUP",
          "languageId": "markdown",
          "metadata": {},
          "outputs": [],
          "value": "down the only road I have ever known",
        },
      },
    },
    "sequence": 11,
  },
]
`;
